_posts/2024-10-28-using-maven-wildfly-plugin-with-postgresql-ds.adoc

```diff
@@ -0,0 +1,195 @@
+
+[source,xml]
+----
+<packaging-scripts>
+    <packaging-script>
+        <scripts>
+            <script>${configure.ds.script}</script>
+        </scripts>
+        <!-- Expressions resolved during server execution -->
+        <resolve-expressions>false</resolve-expressions>
+    </packaging-script>
+</packaging-scripts>
+----
+
+Here is an example that is showing the above configuration method:
+
+* https://github.com/wildfly/quickstart/blob/main/jaxrs-jwt/pom.xml#L177-L185[quickstart/jaxrs-jwt/pom.xml at main · wildfly/quickstart]
+
+Above is the introduction to the script configuration. Besides the `postgresql-driver` layer, the `wildfly-datasources-galleon-pack` also provide other layers as described in the documentfootnote:[https://github.com/wildfly-extras/wildfly-datasources-galleon-pack/blob/main/doc/postgresql/README.md]:
+
+____
+* `postgresql-default-datasource`: Provision the PostgreSQLDS non xa
+datasource and configures it as the default one. Depends on
+postgresql-datasource layer.
+* `postgresql-datasource`: Provision the PostgreSQLDS non xa
+datasource. Depends on postgresql-driver layer.
+* `postgresql-driver`: Provision the postgresql driver. This layer
+installs the JBoss Modules module org.postgresql.jdbc.
+____
+
+As the document described above, the `postgresql-datasource` and the `postgresql-default-datasource` layers define a `PostgreSQLDS` datasource in the provisioned server by default. In addition, the default `PostgreSQLDS` datasource uses some built-in parameters to connect to the underlying PostgreSQL database. Here are the parameters that must be defined:
+
+* `POSTGRESQL_DATABASE`
+* `POSTGRESQL_PASSWORD`
+* `POSTGRESQL_URL`
+* `POSTGRESQL_USER`
+
+I have provided an example of the usage of the `postgresql-datasource` in this pull request:
+
+* https://github.com/jberet/jberet-examples/pull/12[Provide the usage example of the `postgresql-datasource` layer. by liweinan · Pull Request #12 · jberet/jberet-examples]
+
+The above pull request contains the necessary configuration to use the `postgresql-datasource` layer for example. As the example shown above, these environment variables are configured in the
+`wildfly-maven-plugin`:
+
+[source,xml]
+----
+<env>
+    <POSTGRESQL_DATABASE>batch_db</POSTGRESQL_DATABASE>
+    <POSTGRESQL_PASSWORD>123</POSTGRESQL_PASSWORD>
+    <POSTGRESQL_URL>jdbc:postgresql://localhost:5432/batch_db</POSTGRESQL_URL>
+    <POSTGRESQL_USER>batch_user</POSTGRESQL_USER>
+</env>
+----
+
+And the `postgresql-datasource` layer is used:
+
+[source,xml]
+----
+<layers>
+    <layer>postgresql-datasource</layer>
+</layers>
+----
+
+After the above configuration is used, and if the provisioned server is run by `mvn wildfly:run`, then from the server log output I could see:
+
+[source,txt]
+----
+00:38:19,862 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-5) WFLYJCA0001: Bound data source [java:jboss/datasources/PostgreSQLDS]
```

Comment:
Nit: I don't think the message is quite right:
```
15:06:50,687 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-4) WFLYJCA0001: Bound data source [java:jboss/jsr352/batch_db]
```

^^ is what I see. Note the `java:/jboss/jsr352/batch_db`.

Reply:
@jamezp Do you think I should update the `jberet-examples` project to provide another set of CLI scripts that will use the `PostgreSQLDS` datasource? Or maybe just update the README file of the example?

Reply:
@jamezp I have provided the `postgresql-datasource` usage in this PR: https://github.com/jberet/jberet-examples/pull/14 Could you please help to review it when you have time? Thanks!

Reply:
Yes, they will both show if you use the `<layer>postgresql-datasource</layer>`. However, the `Bound data source [java:jboss/jsr352/batch_db]` is the one we're interested in because that is the one we've configured.

This is what I see when I enable the `postgresql-datasource` layer and run `mvn clean wildfly:dev -Ppostgres` on the example:
```
09:39:31,573 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-6) WFLYJCA0001: Bound data source [java:jboss/jsr352/batch_db]
09:39:31,574 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-7) Queuing requests.
09:39:31,574 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-8) WFLYJCA0001: Bound data source [java:jboss/datasources/PostgreSQLDS]
```

I see both, but the `Bound data source [java:jboss/datasources/PostgreSQLDS]` isn't of interest to our deployment because we don't use it.

Reply:
Thanks for the comment James! I'll update the text accordingly.